File: ios/Sources/PoltioSDK/UI/PoltioOverlayManager.swift
#if canImport(UIKit)
    import UIKit

    /// Transparent overlay window that passes through touches except when interacting with the floating trigger.
    @available(iOSApplicationExtension, unavailable)
    public final class PoltioPassthroughWindow: UIWindow {
        private var lastNotificationTime: TimeInterval = 0

        override public func hitTest(_ point: CGPoint, with event: UIEvent?) -> UIView? {
            guard let hitView = super.hitTest(point, with: event) else {
                return nil
            }
            // Pass touches through if the user tapped on the empty window or root controller container view
            if hitView === self || hitView === rootViewController?.view {
                let now = CACurrentMediaTime()
                if now - lastNotificationTime > 0.1 {
                    lastNotificationTime = now
                    NotificationCenter.default.post(name: PoltioFloatingPillTriggerView.didScrollNotification, object: nil)
                }
                return nil
            }
            return hitView
        }
    }

    /// Dedicated root view controller for the overlay window.
    @available(iOSApplicationExtension, unavailable)
    final class PoltioOverlayRootViewController: UIViewController {
        override func viewDidLoad() {
            super.viewDidLoad()
            view.backgroundColor = .clear
            view.isOpaque = false
        }

        override var shouldAutorotate: Bool {
            true
        }
    }

    /// Manages attaching, presenting, and dismissing native Poltio floating triggers and interactive webviews.
    @available(iOSApplicationExtension, unavailable)
    public final class PoltioOverlayManager {
        public static let shared = PoltioOverlayManager()

        /// Maximum number of times `showTrigger` will retry while waiting for a `UIWindowScene` to
        /// become available, before giving up. Caps retries at 10 * 0.2s = 2s so a host app that never
        /// produces an active scene (e.g. unusual lifecycle, extension context) doesn't loop forever.
        private static let maxShowTriggerRetries = 10

        private var overlayWindow: PoltioPassthroughWindow?
        private var activeTriggerView: (UIView & PoltioTriggerPresentable)?
        private var currentPublicId: String?
        private var currentTriggerType: String?
        private var pendingShowWorkItem: DispatchWorkItem?
        private var showTriggerRetryCount = 0

        private init() {}

        /// Displays the floating trigger for the resolved widget on a dedicated passthrough overlay window.
        /// - Parameters:
        ///   - widget: The resolved Poltio widget configuration.
        ///   - puid: Optional developer-provided user identifier.
        public func showTrigger(
            widget: PoltioWidgetResponse,
            puid: String? = nil
        ) {
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }

                pendingShowWorkItem?.cancel()
                pendingShowWorkItem = nil

                guard widget.overlayOptions.isBoxTrigger || widget.overlayOptions.isPillTrigger || widget.overlayOptions.isCardTrigger else {
                    PoltioLogger.warning("Trigger type '\(widget.overlayOptions.triggerType ?? "none")' is not currently handled (supported: 'box', 'pill', 'card').")
                    hideTrigger()
                    return
                }

                guard !widget.overlayOptions.hideButton else {
                    PoltioLogger.debug("Floating trigger suppressed for widget '\(widget.publicId)' (floating-hide-button).")
                    hideTrigger()
                    return
                }

                guard !PoltioTriggerDismissalStore.isDismissed(publicId: widget.publicId) else {
                    PoltioLogger.debug("Floating trigger suppressed for widget '\(widget.publicId)' (still within its close-remember window).")
                    hideTrigger()
                    return
                }

                let targetTriggerType = widget.overlayOptions.triggerType ?? ""

                if currentPublicId == widget.publicId,
                   currentTriggerType == targetTriggerType,
                   activeTriggerView != nil,
                   activeTriggerView?.superview != nil
                {
                    // Already active and visible for this exact widget and trigger type
                    return
                }

                // A retry (scheduled by this same method below) re-enters here for the same widget while
                // it's still waiting on a UIWindowScene. Only reset the retry counter for a genuinely new
                // widget — otherwise it gets zeroed on every retry and the cap below never trips.
                let isRetryForSameWidget = currentPublicId == widget.publicId

                // Clean up previous overlay and update state immediately
                teardownOverlaySynchronously()
                if !isRetryForSameWidget {
                    showTriggerRetryCount = 0
                }
                currentPublicId = widget.publicId
                currentTriggerType = targetTriggerType

                guard let windowScene = findActiveWindowScene() else {
                    guard showTriggerRetryCount < Self.maxShowTriggerRetries else {
                        PoltioLogger.warning("UIWindowScene still not available after \(Self.maxShowTriggerRetries) retries; giving up on showing trigger for widget '\(widget.publicId)'.")
                        showTriggerRetryCount = 0
                        currentPublicId = nil
                        currentTriggerType = nil
                        return
                    }
                    showTriggerRetryCount += 1
                    PoltioLogger.debug("UIWindowScene not ready yet, retrying showTrigger (\(showTriggerRetryCount)/\(Self.maxShowTriggerRetries)) after 0.2s...")
                    let workItem = DispatchWorkItem { [weak self] in
                        guard let self, currentPublicId == widget.publicId else { return }
                        showTrigger(widget: widget, puid: puid)
                    }
                    pendingShowWorkItem = workItem
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.2, execute: workItem)
                    return
                }

                showTriggerRetryCount = 0

                let window: PoltioPassthroughWindow
                if #available(iOS 13.0, *) {
                    window = PoltioPassthroughWindow(windowScene: windowScene)
                    window.frame = windowScene.coordinateSpace.bounds
                } else {
                    window = PoltioPassthroughWindow(frame: UIScreen.main.bounds)
                }

                // `floatingZindex` (default 100) nudges the overlay window's level around its base position
                // instead of remapping it wholesale, so integrators can order it relative to other custom
                // overlay windows without risking it jumping above system alerts.
                let zIndexOffset = max(-0.5, min(0.5, CGFloat(widget.overlayOptions.floatingZindex - 100) * 0.01))
                window.windowLevel = UIWindow.Level(rawValue: UIWindow.Level.alert.rawValue - 1 + zIndexOffset)
                window.backgroundColor = .clear
                window.isOpaque = false

                let rootVC = PoltioOverlayRootViewController()
                window.rootViewController = rootVC

                let onOpenWidget: () -> Void = { [weak self] in
                    self?.presentWidgetWebView(publicId: widget.publicId, puid: puid, overlayOptions: widget.overlayOptions)
                }
                let onDismissForever: (Double) -> Void = { [weak self] hours in
                    PoltioTriggerDismissalStore.recordDismissal(publicId: widget.publicId, hours: hours)
                    self?.hideTrigger()
                }

                let vertical = widget.overlayOptions.verticalPosition
                let horizontal = widget.overlayOptions.horizontalPosition
                let triggerView: UIView & PoltioTriggerPresentable

                if widget.overlayOptions.isPillTrigger {
                    let pillView = PoltioFloatingPillTriggerView(
                        widget: widget,
                        onOpenWidget: onOpenWidget,
                        onDismissForever: onDismissForever
                    )
                    pillView.translatesAutoresizingMaskIntoConstraints = false
                    rootVC.view.addSubview(pillView)
                    pinTrigger(
                        pillView, in: rootVC.view, vertical: vertical, horizontal: horizontal,
                        horizontalInset: 16, verticalInset: 70, hugsRawEdgeHorizontally: false
                    )
                    triggerView = pillView
                } else if widget.overlayOptions.isCardTrigger {
                    let cardView = PoltioFloatingCardTriggerView(
                        widget: widget,
                        onOpenWidget: onOpenWidget
                    )
                    cardView.translatesAutoresizingMaskIntoConstraints = false
                    rootVC.view.addSubview(cardView)
                    pinTrigger(
                        cardView, in: rootVC.view, vertical: vertical, horizontal: horizontal,
                        horizontalInset: 0, verticalInset: 70, hugsRawEdgeHorizontally: true
                    )
                    triggerView = cardView
                } else {
                    let boxView = PoltioFloatingBoxTriggerView(
                        widget: widget,
                        onOpenWidget: onOpenWidget,
                        onDismissForever: onDismissForever
                    )
                    boxView.translatesAutoresizingMaskIntoConstraints = false
                    rootVC.view.addSubview(boxView)
                    pinTrigger(
                        boxView, in: rootVC.view, vertical: vertical, horizontal: horizontal,
                        horizontalInset: 0, verticalInset: 24, hugsRawEdgeHorizontally: true
                    )
                    triggerView = boxView
                }

                activeTriggerView = triggerView
                overlayWindow = window
                window.isHidden = false

                PoltioLogger.info("Attached floating trigger overlay for widget '\(widget.publicId)' (type: \(widget.overlayOptions.triggerType ?? "unknown")) in dedicated passthrough window.")

                triggerView.alpha = 0
                triggerView.transform = CGAffineTransform(translationX: 80, y: 0)

                UIView.animate(
                    withDuration: 0.4,
                    delay: 0.05,
                    usingSpringWithDamping: 0.8,
                    initialSpringVelocity: 0.5,
                    options: [.curveEaseOut, .allowUserInteraction],
                    animations: {
                        triggerView.alpha = 1
                        triggerView.transform = .identity
                    },
                    completion: nil
                )
            }
        }

        /// Hides and removes any currently displayed floating trigger view and its overlay window with animation.
        public func hideTrigger() {
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                pendingShowWorkItem?.cancel()
                pendingShowWorkItem = nil
                currentPublicId = nil
                currentTriggerType = nil
                let windowToClose = overlayWindow
                let viewToClose = activeTriggerView

                overlayWindow = nil
                activeTriggerView = nil

                guard let view = viewToClose, let window = windowToClose else { return }

                UIView.animate(withDuration: 0.25, animations: {
                    view.alpha = 0
                    view.transform = CGAffineTransform(translationX: 80, y: 0)
                }, completion: { _ in
                    view.removeFromSuperview()
                    window.isHidden = true
                })
            }
        }

        /// Synchronously tears down any existing overlay window/view.
        private func teardownOverlaySynchronously() {
            pendingShowWorkItem?.cancel()
            pendingShowWorkItem = nil
            currentPublicId = nil
            currentTriggerType = nil
            activeTriggerView?.removeFromSuperview()
            activeTriggerView = nil
            overlayWindow?.isHidden = true
            overlayWindow = nil
        }

        /// Pins `view` within `container` according to the parsed `floatingPosition` (`vertical`/`horizontal`),
        /// generalizing what used to be three hardcoded per-trigger constraint blocks. `hugsRawEdgeHorizontally`
        /// preserves each trigger's pre-existing horizontal-anchoring convention (card/box hug the raw view
        /// edge so their collapsed tabs bleed off-screen; pill insets from the safe area).
        private func pinTrigger(
            _ view: UIView,
            in container: UIView,
            vertical: String,
            horizontal: String,
            horizontalInset: CGFloat,
            verticalInset: CGFloat,
            hugsRawEdgeHorizontally: Bool
        ) {
            var constraints: [NSLayoutConstraint] = []

            if horizontal == "left" {
                let anchor = hugsRawEdgeHorizontally ? container.leadingAnchor : container.safeAreaLayoutGuide.leadingAnchor
                constraints.append(view.leadingAnchor.constraint(equalTo: anchor, constant: horizontalInset))
            } else {
                let anchor = hugsRawEdgeHorizontally ? container.trailingAnchor : container.safeAreaLayoutGuide.trailingAnchor
                constraints.append(view.trailingAnchor.constraint(equalTo: anchor, constant: -horizontalInset))
            }

            switch vertical {
            case "top":
                constraints.append(view.topAnchor.constraint(equalTo: container.safeAreaLayoutGuide.topAnchor, constant: verticalInset))
            case "center":
                constraints.append(view.centerYAnchor.constraint(equalTo: container.centerYAnchor, constant: -verticalInset))
            default: // "bottom"
                constraints.append(view.bottomAnchor.constraint(equalTo: container.safeAreaLayoutGuide.bottomAnchor, constant: -verticalInset))
            }

            NSLayoutConstraint.activate(constraints)
        }

        /// Presents the interactive widget modal WebView on top of the active view controller.
        public func presentWidgetWebView(publicId: String, puid: String?, overlayOptions: PoltioOverlayOptions? = nil) {
            DispatchQueue.main.async { [weak self] in
                guard let self, let topVC = findTopmostHostViewController() else {
                    PoltioLogger.error("Unable to find topmost view controller to present widget.")
                    return
                }

                // Hide the floating trigger window while the webview is presented
                overlayWindow?.isHidden = true

                let webVC = PoltioWebViewController(publicId: publicId, puid: puid, overlayOptions: overlayOptions)
                webVC.onWidgetEvent = { event, data in
                    PoltioSDK.onWidgetEvent?(event, data)
                }
                webVC.onDismiss = { [weak self] in
                    DispatchQueue.main.async {
                        guard let self, let window = self.overlayWindow, let trigger = self.activeTriggerView else {
                            return
                        }

                        // Reset to collapsed tab after user interacted with webview
                        trigger.resetToCollapsed(animated: false)

                        window.isHidden = false
                        trigger.alpha = 0
                        trigger.transform = CGAffineTransform(translationX: 80, y: 0)

                        UIView.animate(
                            withDuration: 0.35,
                            delay: 0.05,
                            usingSpringWithDamping: 0.8,
                            initialSpringVelocity: 0.5,
                            options: [.curveEaseOut, .allowUserInteraction],
                            animations: {
                                trigger.alpha = 1
                                trigger.transform = .identity
                            },
                            completion: nil
                        )
                    }
                }

                topVC.present(webVC, animated: true, completion: nil)
            }
        }

        // MARK: - Scene & View Controller Traversal

        private func findActiveWindowScene() -> UIWindowScene? {
            if #available(iOS 13.0, *) {
                let scenes = UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }

                if let active = scenes.first(where: { $0.activationState == .foregroundActive }) {
                    return active
                }
                if let first = scenes.first {
                    return first
                }
                if let scene = UIApplication.shared.windows.first?.windowScene {
                    return scene
                }
            }
            return nil
        }

        private func findHostKeyWindow() -> UIWindow? {
            if #available(iOS 13.0, *) {
                let scenes = UIApplication.shared.connectedScenes
                    .compactMap { $0 as? UIWindowScene }

                for scene in scenes {
                    if let window = scene.windows.first(where: { $0 !== self.overlayWindow && $0.isKeyWindow && $0.rootViewController != nil }) {
                        return window
                    }
                    if let window = scene.windows.first(where: { $0 !== self.overlayWindow && $0.rootViewController != nil }) {
                        return window
                    }
                    if let window = scene.windows.first(where: { $0 !== self.overlayWindow }) {
                        return window
                    }
                }
            }

            return UIApplication.shared.windows.first(where: { $0 !== self.overlayWindow && $0.isKeyWindow && $0.rootViewController != nil })
                ?? UIApplication.shared.windows.first(where: { $0 !== self.overlayWindow && $0.rootViewController != nil })
                ?? UIApplication.shared.windows.first(where: { $0 !== self.overlayWindow })
                ?? UIApplication.shared.keyWindow
        }

        private func findTopmostHostViewController(from root: UIViewController? = nil) -> UIViewController? {
            let base = root ?? findHostKeyWindow()?.rootViewController

            if let nav = base as? UINavigationController {
                return findTopmostHostViewController(from: nav.visibleViewController)
            }
            if let tab = base as? UITabBarController {
                return findTopmostHostViewController(from: tab.selectedViewController)
            }
            if let presented = base?.presentedViewController {
                return findTopmostHostViewController(from: presented)
            }

            return base
        }
    }
#endif

File: android/poltio-sdk/src/main/java/com/poltio/sdk/ui/PoltioOverlayManager.kt
package com.poltio.sdk.ui

import android.app.Activity
import android.app.Application
import android.content.Intent
import android.os.Bundle
import android.view.Gravity
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.FrameLayout
import com.poltio.sdk.PoltioExecutors
import com.poltio.sdk.PoltioLogger
import com.poltio.sdk.PoltioTriggerDismissalStore
import com.poltio.sdk.PoltioWidgetResponse
import java.lang.ref.WeakReference

/**
 * Manages attaching, presenting, and dismissing native Poltio floating triggers and the
 * interactive widget WebView.
 *
 * Unlike iOS (which needs a dedicated passthrough `UIWindow` to overlay across the whole app),
 * Android's single-window view hierarchy lets a sibling `FrameLayout` added on top of the current
 * Activity's content view achieve the same click-through effect for free: a touch that doesn't
 * land on the trigger simply isn't consumed, so the framework's own dispatch falls through to the
 * content view beneath it. This mirrors AGENTS.md's guidance to attach to the current Activity's
 * `window.decorView` / root `FrameLayout` rather than a system-alert-window overlay.
 */
internal object PoltioOverlayManager {
    private const val MAX_SHOW_RETRIES = 10
    private const val RETRY_DELAY_MS = 200L

    /**
     * Default bottom inset for the box trigger — unlike pill/card (70dp), the box trigger's
     * collapsed tab is tall (140dp) and sits low by design, but 24dp put it directly behind most
     * apps' bottom navigation bars (Material's default `NavigationBar` height is 80dp). This
     * clears that by default across screen sizes, since dp — unlike raw pixels — already scales
     * with density rather than physical screen size; [avoidSystemBars] on top of it additionally
     * accounts for the real gesture/3-button system nav bar at runtime.
     */
    private const val BOX_DEFAULT_BOTTOM_INSET_DP = 88f

    private var application: Application? = null
    private var resumedActivity: WeakReference<Activity>? = null

    private var overlayContainer: FrameLayout? = null
    private var activeTriggerView: (android.view.View)? = null
    private var currentPublicId: String? = null
    private var currentTriggerType: String? = null
    private var showTriggerRetryCount = 0
    private var pendingShowRunnable: Runnable? = null

    /** Registers activity-lifecycle tracking; safe to call more than once with the same instance. */
    fun attach(app: Application) {
        if (application === app) return
        application = app
        app.registerActivityLifecycleCallbacks(object : Application.ActivityLifecycleCallbacks {
            override fun onActivityResumed(activity: Activity) {
                resumedActivity = WeakReference(activity)
            }

            override fun onActivityPaused(activity: Activity) {
                if (resumedActivity?.get() === activity) resumedActivity = null
            }

            override fun onActivityDestroyed(activity: Activity) {
                if (resumedActivity?.get() === activity) resumedActivity = null
                if (overlayContainer?.let { it.context === activity } == true) {
                    teardownOverlaySynchronously()
                }
            }

            override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
            override fun onActivityStarted(activity: Activity) = Unit
            override fun onActivityStopped(activity: Activity) = Unit
            override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
        })
    }

    /** Displays the floating trigger for the resolved widget on the current Activity, on the main thread. */
    fun showTrigger(widget: PoltioWidgetResponse, puid: String?) {
        PoltioExecutors.runOnMain { showTriggerOnMain(widget, puid) }
    }

    /** Hides and removes any currently displayed floating trigger, on the main thread. */
    fun hideTrigger() {
        PoltioExecutors.runOnMain { hideTriggerOnMain() }
    }

    private fun showTriggerOnMain(widget: PoltioWidgetResponse, puid: String?) {
        pendingShowRunnable?.let { PoltioExecutors.main.removeCallbacks(it) }
        pendingShowRunnable = null

        val options = widget.overlayOptions
        if (!(options.isBoxTrigger || options.isPillTrigger || options.isCardTrigger)) {
            PoltioLogger.warning { "Trigger type '${options.triggerType ?: "none"}' is not currently handled (supported: 'box', 'pill', 'card')." }
            hideTriggerOnMain()
            return
        }

        if (options.hideButton) {
            PoltioLogger.debug { "Floating trigger suppressed for widget '${widget.publicId}' (floating-hide-button)." }
            hideTriggerOnMain()
            return
        }

        val activity = resumedActivity?.get()
        if (activity != null && PoltioTriggerDismissalStore.isDismissed(activity, widget.publicId)) {
            PoltioLogger.debug { "Floating trigger suppressed for widget '${widget.publicId}' (still within its close-remember window)." }
            hideTriggerOnMain()
            return
        }

        val targetTriggerType = options.triggerType ?: ""

        if (currentPublicId == widget.publicId &&
            currentTriggerType == targetTriggerType &&
            activeTriggerView != null &&
            activeTriggerView?.parent != null
        ) {
            // Already active and visible for this exact widget and trigger type.
            return
        }

        val isRetryForSameWidget = currentPublicId == widget.publicId
        teardownOverlaySynchronously()
        if (!isRetryForSameWidget) showTriggerRetryCount = 0
        currentPublicId = widget.publicId
        currentTriggerType = targetTriggerType

        if (activity == null) {
            if (showTriggerRetryCount >= MAX_SHOW_RETRIES) {
                PoltioLogger.warning { "No resumed Activity available after $MAX_SHOW_RETRIES retries; giving up on showing trigger for widget '${widget.publicId}'." }
                showTriggerRetryCount = 0
                currentPublicId = null
                currentTriggerType = null
                return
            }
            showTriggerRetryCount++
            PoltioLogger.debug { "No resumed Activity yet, retrying showTrigger ($showTriggerRetryCount/$MAX_SHOW_RETRIES) after ${RETRY_DELAY_MS}ms..." }
            val runnable = Runnable {
                if (currentPublicId == widget.publicId) showTriggerOnMain(widget, puid)
            }
            pendingShowRunnable = runnable
            PoltioExecutors.main.postDelayed(runnable, RETRY_DELAY_MS)
            return
        }

        showTriggerRetryCount = 0

        val contentRoot = activity.findViewById<ViewGroup>(android.R.id.content) ?: return
        val container = FrameLayout(activity).apply {
            layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
            clipChildren = false
            clipToPadding = false
        }
        installOutsideTouchDetector(container)
        contentRoot.addView(container)
        overlayContainer = container

        // Android has no direct analogue of a CSS/window z-index; `floating-zindex` is instead
        // mapped onto view elevation so a host app that stacks its own elevated views (dialogs,
        // app bars) can still be told to render the trigger more/less prominently above them.
        // Clamped to Material's practical elevation range so an arbitrary large value doesn't
        // produce a runaway shadow.
        val zIndexElevationDp = (options.floatingZindex / 100.0 * 8.0).coerceIn(0.0, 24.0)
        androidx.core.view.ViewCompat.setElevation(container, activity.dp(zIndexElevationDp.toFloat()).toFloat())

        val onOpenWidget: () -> Unit = { presentWidgetWebView(widget.publicId, puid, widget.overlayOptions) }
        val onDismissForever: (Double) -> Unit = { hours ->
            resumedActivity?.get()?.let { PoltioTriggerDismissalStore.recordDismissal(it, widget.publicId, hours) }
            hideTriggerOnMain()
        }

        val vertical = options.verticalPosition
        val horizontal = options.horizontalPosition

        val triggerView: android.view.View = when {
            options.isPillTrigger -> PoltioFloatingPillTriggerView(activity, widget, onOpenWidget, onDismissForever).also {
                pinTrigger(it, container, vertical, horizontal, horizontalInsetDp = 16f, verticalInsetDp = 70f, avoidSystemBars = true)
            }
            options.isCardTrigger -> PoltioFloatingCardTriggerView(activity, widget, onOpenWidget).also {
                pinTrigger(it, container, vertical, horizontal, horizontalInsetDp = 0f, verticalInsetDp = 70f, avoidSystemBars = false)
            }
            else -> PoltioFloatingBoxTriggerView(activity, widget, onOpenWidget, onDismissForever).also {
                pinTrigger(it, container, vertical, horizontal, horizontalInsetDp = 0f, verticalInsetDp = BOX_DEFAULT_BOTTOM_INSET_DP, avoidSystemBars = true)
            }
        }

        activeTriggerView = triggerView

        PoltioLogger.info { "Attached floating trigger overlay for widget '${widget.publicId}' (type: ${options.triggerType ?: "unknown"})." }

        animateTriggerIn(triggerView)
    }

    private fun animateTriggerIn(view: android.view.View) {
        val restingX = view.translationX
        view.alpha = 0f
        view.translationX = restingX + view.context.dp(80f)
        view.animate()
            .alpha(1f)
            .translationX(restingX)
            .setStartDelay(50)
            .setDuration(400)
            .setInterpolator(android.view.animation.OvershootInterpolator(0.9f))
            .start()
    }

    private fun hideTriggerOnMain() {
        pendingShowRunnable?.let { PoltioExecutors.main.removeCallbacks(it) }
        pendingShowRunnable = null
        currentPublicId = null
        currentTriggerType = null

        val container = overlayContainer
        val view = activeTriggerView
        overlayContainer = null
        activeTriggerView = null

        if (view == null || container == null) return

        view.animate()
            .alpha(0f)
            .translationX(view.translationX + view.context.dp(80f))
            .setDuration(250)
            .withEndAction {
                (container.parent as? ViewGroup)?.removeView(container)
            }
            .start()
    }

    /** Synchronously tears down any existing overlay container/view (no animation). */
    private fun teardownOverlaySynchronously() {
        pendingShowRunnable?.let { PoltioExecutors.main.removeCallbacks(it) }
        pendingShowRunnable = null
        currentPublicId = null
        currentTriggerType = null
        activeTriggerView = null
        overlayContainer?.let { (it.parent as? ViewGroup)?.removeView(it) }
        overlayContainer = null
    }

    /**
     * Detects touches that land on the overlay container but not on any trigger child, and
     * broadcasts them via [PoltioHostInteractionBus] — the Android analogue of iOS's
     * `PoltioPassthroughWindow.hitTest` scroll/tap-outside notification. The overlay never
     * consumes these events, so they still fall through to the host screen underneath.
     */
    private fun installOutsideTouchDetector(container: FrameLayout) {
        container.setOnTouchListener { _, event ->
            if (event.actionMasked == MotionEvent.ACTION_DOWN) {
                PoltioHostInteractionBus.notifyOutsideInteraction()
            }
            false
        }
    }

    /**
     * Pins [view] within [container] according to the parsed `floatingPosition`
     * (`vertical`/`horizontal`), applying system-bar insets only where iOS relied on the safe-area
     * layout guide (the pill trigger); the card/box tabs intentionally hug the raw edge so their
     * collapsed tab bleeds off-screen, matching iOS.
     */
    private fun pinTrigger(
        view: android.view.View,
        container: FrameLayout,
        vertical: String,
        horizontal: String,
        horizontalInsetDp: Float,
        verticalInsetDp: Float,
        avoidSystemBars: Boolean,
    ) {
        val context = container.context
        val params = view.layoutParams as? FrameLayout.LayoutParams
            ?: FrameLayout.LayoutParams(FrameLayout.LayoutParams.WRAP_CONTENT, FrameLayout.LayoutParams.WRAP_CONTENT)

        val horizontalGravity = if (horizontal == "left") Gravity.START else Gravity.END
        val verticalGravity = when (vertical) {
            "top" -> Gravity.TOP
            "center" -> Gravity.CENTER_VERTICAL
            else -> Gravity.BOTTOM
        }
        params.gravity = horizontalGravity or verticalGravity

        val horizontalInsetPx = context.dp(horizontalInsetDp)
        val verticalInsetPx = context.dp(verticalInsetDp)

        fun applyMargins(systemLeft: Int, systemRight: Int, systemTop: Int, systemBottom: Int) {
            if (horizontal == "left") params.leftMargin = horizontalInsetPx + systemLeft else params.rightMargin = horizontalInsetPx + systemRight
            when (vertical) {
                "top" -> params.topMargin = verticalInsetPx + systemTop
                "center" -> Unit
                else -> params.bottomMargin = verticalInsetPx + systemBottom
            }
            view.layoutParams = params
        }

        // Position with zero system-bar compensation up front so the trigger doesn't wait on a
        // window-insets dispatch to appear at all.
        applyMargins(systemLeft = 0, systemRight = 0, systemTop = 0, systemBottom = 0)
        container.addView(view, params)
        if (vertical == "center") view.translationY = -verticalInsetPx.toFloat()

        if (avoidSystemBars) {
            // `ViewCompat.getRootWindowInsets(container)` right after `addView` is unreliable —
            // insets aren't dispatched until the next layout pass, so a synchronous read here can
            // silently return null. `setOnApplyWindowInsetsListener` instead reacts whenever insets
            // actually become available (and again on later changes, e.g. rotation).
            androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener(container) { _, insets ->
                val systemBars = insets.getInsets(androidx.core.view.WindowInsetsCompat.Type.systemBars())
                applyMargins(systemBars.left, systemBars.right, systemBars.top, systemBars.bottom)
                insets
            }
        }
    }

    /** Presents the interactive widget modal WebView on top of the resumed Activity. */
    fun presentWidgetWebView(publicId: String, puid: String?, overlayOptions: com.poltio.sdk.PoltioOverlayOptions? = null) {
        PoltioExecutors.runOnMain {
            val activity = resumedActivity?.get()
            if (activity == null) {
                PoltioLogger.error { "Unable to find a resumed Activity to present the widget." }
                return@runOnMain
            }

            // Hide (but don't tear down) the floating trigger while the webview is presented.
            overlayContainer?.visibility = android.view.View.INVISIBLE

            val intent = PoltioWebViewActivity.newIntent(activity, publicId, puid, overlayOptions)
            activity.startActivity(intent)
            // PoltioWebViewActivity drives its own slide-up/scrim-fade entrance animation against
            // its translucent theme; suppress the system's default activity-enter transition so
            // the two don't stack.
            activity.overridePendingTransition(0, 0)
        }
    }

    /** Called by [PoltioWebViewActivity] when it's dismissed, to restore the collapsed trigger. */
    internal fun onWidgetWebViewDismissed() {
        PoltioExecutors.runOnMain {
            val container = overlayContainer ?: return@runOnMain
            val trigger = activeTriggerView as? PoltioTriggerPresentable

            trigger?.resetToCollapsed(animated = false)
            container.visibility = android.view.View.VISIBLE

            val view = activeTriggerView ?: return@runOnMain
            animateTriggerIn(view)
        }
    }
}
